genericjpa/common-apis/src/main/java/org/hibernate/search/genericjpa/metadata/impl/RehashedTypeMetadata.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.genericjpa.metadata.impl;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.hibernate.annotations.common.reflection.XProperty;
+import org.hibernate.search.backend.spi.SingularTermDeletionQuery;
+import org.hibernate.search.engine.metadata.impl.DocumentFieldMetadata;
+import org.hibernate.search.engine.metadata.impl.TypeMetadata;
+
+/**
+ * @author Martin Braun
+ */
+public final class RehashedTypeMetadata {
```

Comment:
What's the purpose of this class? There should always be a meaningful class-level comment. Also some more high-level documentation is needed (e.g. package-level JavaDocs) to describe the overall mechanism. That's the first thing I'd really need to grasp what's going on here :) I.e. no need for polished JavaDocs everywhere, but something providing an overview would help.